scipy/signal/signaltools.py

```diff
@@ -2585,12 +2507,117 @@ def residuez(b, a, tol=1e-3, rtype='avg'):
 
     residues *= (-poles) ** powers / a_rev[0]
 
-    return residues, poles, np.trim_zeros(k_rev[::-1], 'b')
+    return residues, poles, k_rev[::-1]
 
 
-def invresz(r, p, k, tol=1e-3, rtype='avg'):
+def _group_poles(poles, tol, rtype):
+    if rtype in ['max', 'maximum']:
+        reduce = np.max
+    elif rtype in ['min', 'minimum']:
+        reduce = np.min
+    elif rtype in ['avg', 'mean']:
+        reduce = np.mean
+    else:
+        raise ValueError("`rtype` must be one of "
+                         "{'max', 'maximum', 'min', 'minimum', 'avg', 'mean'}")
+
+    unique = []
+    multiplicity = []
+
+    pole = poles[0]
+    block = [pole]
+    for i in range(1, len(poles)):
+        if abs(poles[i] - pole) <= tol:
+            block.append(pole)
+        else:
+            unique.append(reduce(block))
+            multiplicity.append(len(block))
+            pole = poles[i]
+            block = [pole]
+
+    unique.append(reduce(block))
+    multiplicity.append(len(block))
+
+    return np.asarray(unique), np.asarray(multiplicity)
+
+
+def invres(r, p, k, tol=1e-3, rtype='avg'):
```

Comment:
Can you move this back up to where it was before? Currently the diff is much larger than it needs to be because this moved down in the file, which makes reviewing more difficult.

Reply:
Done.

I suggest that we move all partial fraction expansion code into a separate file in a follow-up PR, because `signaltools.py` is too lengthy.